handler-ssl-ocsp/src/main/java/io/netty/handler/ssl/ocsp/OcspServerCertificateValidator.java

```diff
@@ -0,0 +1,142 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl.ocsp;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.ssl.SslHandler;
+import io.netty.handler.ssl.SslHandshakeCompletionEvent;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.GenericFutureListener;
+import io.netty.util.concurrent.Promise;
+import org.bouncycastle.cert.ocsp.BasicOCSPResp;
+import org.bouncycastle.cert.ocsp.CertificateStatus;
+import org.bouncycastle.cert.ocsp.OCSPException;
+import org.bouncycastle.cert.ocsp.RevokedStatus;
+import org.bouncycastle.cert.ocsp.SingleResp;
+
+import java.security.cert.Certificate;
+import java.security.cert.X509Certificate;
+
+/**
+ * {@link OcspServerCertificateValidator} validates incoming server's certificate
+ * using OCSP. Once TLS handshake is completed, {@link SslHandshakeCompletionEvent#SUCCESS} is fired, validator
+ * will perform certificate validation using OCSP over HTTP/1.1 with the server's certificate issuer OCSP responder.
+ */
+public class OcspServerCertificateValidator extends ChannelInboundHandlerAdapter {
+
+    private final boolean validateNonce;
+    private final ConnectionChannelFactory channelFactory;
+
+    /**
+     * Create a new {@link OcspServerCertificateValidator} instance without nonce validation
+     * on OCSP response and using default {@link ConnectionChannelFactory} instance.
+     */
+    public OcspServerCertificateValidator() {
+        this(false);
+    }
+
+    /**
+     * Create a new {@link OcspServerCertificateValidator} instance with
+     * default {@link ConnectionChannelFactory} implementation.
+     *
+     * @param validateNonce Set to {@link Boolean#TRUE} if we should force nonce validation on
+     *                      OCSP response else set to {@link Boolean#FALSE}
+     */
+    public OcspServerCertificateValidator(boolean validateNonce) {
+        this(validateNonce, null);
+    }
+
+    /**
+     * Create a new {@link OcspServerCertificateValidator} instance
+     *
+     * @param validateNonce  Set to {@link Boolean#TRUE} if we should force nonce validation on
+     *                       OCSP response else set to {@link Boolean#FALSE}
+     * @param channelFactory Use {@link ConnectionChannelFactory} instance
+     */
+    public OcspServerCertificateValidator(boolean validateNonce, ConnectionChannelFactory channelFactory) {
+        this.validateNonce = validateNonce;
+
+        if (channelFactory == null) {
+            this.channelFactory = DefaultConnectionChannelFactory.INSTANCE;
```

Comment:
By default I think it would be good to pick the event loop and channel type from the `ChannelHandlerContext` in `userEventTriggered`, instead of making up-front choices about those here.

Reply:
It's a good idea but we cannot use custom transport with this approach. Like I use a DPDK-based one in Netty.